scheduler/propertyset.go

```diff
@@ -171,27 +201,40 @@ func (p *propertySet) SatisfiesDistinctProperties(option *structs.Node, tg strin
 		return false, fmt.Sprintf("missing property %q", p.constraint.LTarget)
 	}
 
-	// both is used to iterate over both the proposed and existing used
-	// properties
-	bothAll := []map[string]struct{}{p.existingValues, p.proposedValues}
-
-	// Check if the nodes value has already been used.
-	for _, usedProperties := range bothAll {
-		// Check if the nodes value has been used
-		_, used := usedProperties[nValue]
-		if !used {
-			continue
+	// combine the counts of how many times the property has been used by
+	// existing and proposed allocations
+	combinedUse := make(map[string]uint64, helper.IntMax(len(p.existingValues), len(p.proposedValues)))
+	for _, usedValues := range []map[string]uint64{p.existingValues, p.proposedValues} {
+		for propertyValue, usedCount := range usedValues {
+			combinedUse[propertyValue] += usedCount
 		}
+	}
 
-		// Check if the value has been cleared from a proposed stop
-		if _, cleared := p.clearedValues[nValue]; cleared {
+	// Go through and discount the combined count when the value has been
+	// cleared by a proposed stop.
+	for propertyValue, clearedCount := range p.clearedValues {
+		combined, ok := combinedUse[propertyValue]
+		if !ok {
 			continue
 		}
 
-		return false, fmt.Sprintf("distinct_property: %s=%s already used", p.constraint.LTarget, nValue)
+		// Don't clear below 0.
+		combinedUse[propertyValue] = helper.Uint64Max(0, combined-clearedCount)
```

Comment:
Since you're using uint if `combined < clearedCount`, the value will wrap and be extremely large instead of getting clamped to 0.

I think this should be:
```go
if combined >= clearedCount {
  combinedUse[propertyValue] = combined-clearedCount
} else {
  combinedUse[propertyValue] = 0
}
```